File: proyecto/cppcalc/ast.cpp
#include "ast.h"
#include <iostream>
#include "calculator.h"


UnaryNode::UnaryNode(AST *sub) :
        AST(),
        subTree(sub) { }

UnaryNode::~UnaryNode() {
#ifdef debug
    cout << "In UnaryNode destructor" << endl;
#endif

    try {
        delete subTree;
    } catch (...) { }
}

AST *UnaryNode::getSubTree() const {
    return subTree;
}

AddNode::AddNode(AST *left, AST *right) :
        BinaryNode(left, right) { }

int AddNode::evaluate() {
    return getLeftSubTree()->evaluate() + getRightSubTree()->evaluate();
}

SubNode::SubNode(AST *left, AST *right) :
        BinaryNode(left, right) { }

int SubNode::evaluate() {
    return getLeftSubTree()->evaluate() - getRightSubTree()->evaluate();
}

NumNode::NumNode(int n) :
        AST(),
        val(n) { }

int NumNode::evaluate() {
    return val;
}

StoreNode::StoreNode(AST *sub) : UnaryNode(sub) { }

int StoreNode::evaluate() {
    int result = getSubTree()->evaluate();
    calc->store(result);
    return result;
}

DivideNode::DivideNode(AST *left, AST *right) :
        BinaryNode(left, right) { }

int DivideNode::evaluate() {
    return getLeftSubTree()->evaluate() / getRightSubTree()->evaluate();
}

TimesNode::TimesNode(AST *left, AST *right) :
        BinaryNode(left, right) { }

int TimesNode::evaluate() {
    return getLeftSubTree()->evaluate() * getRightSubTree()->evaluate();
}

RecallNode::RecallNode() { }

int RecallNode::evaluate() {
    return calc->recall();
}

ModNode::ModNode(AST *left, AST *right) : BinaryNode(left, right){ }

MemPlusNode::MemPlusNode(AST *sub) : UnaryNode(sub) {}

int MemPlusNode::evaluate() {
    int result = getSubTree()->evaluate();
    result = calc->plus(result);
    return result;
}

MemMinusNode::MemMinusNode(AST *sub) : UnaryNode(sub) {}

int MemMinusNode::evaluate() {
    int result = getSubTree()->evaluate();
    result = calc->minus(result);
    return result;
}

int ModNode::evaluate() {
    return  getLeftSubTree()->evaluate() % getRightSubTree()->evaluate();
}

ClearNode::ClearNode() {}

int ClearNode::evaluate() {
    calc->clear();
    return 0;
}

AssignNode::AssignNode(AST_<std::string> *left, AST *right) : BinaryNode_<string, int>(left, right){ }

int AssignNode::evaluate(){
    int result = getRightSubTree()->evaluate();
    std::string name = getLeftSubTree()->evaluate();
    calc->storevar(name, result);
    calc->setAssignPerformed(true);
    cout << "=> " <<  name << " <- ";
    return result;
}

IdentifierNode::IdentifierNode(std::string str) : AST_<std::string>(), val(str){ }

std::string IdentifierNode::evaluate(){
    return val;
}

File: proyecto/rubycalc/lib/ast.rb
require 'set'

class BinaryNode
  attr_reader :left, :right

  def initialize(left, right)
    @left = left
    @right = right
  end
end

class UnaryNode
  attr_reader :sub_tree

  def initialize(sub_tree)
    @sub_tree = sub_tree
  end
end

#NODOS BINARIOS

class AddNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    @left.evaluate + @right.evaluate
  end
end

class SubNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    @left.evaluate - @right.evaluate
  end
end

class TimesNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    @left.evaluate * @right.evaluate
  end
end

class DivideNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    @left.evaluate / @right.evaluate
  end
end

class ModulusNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    @left.evaluate % @right.evaluate
  end
end

class AssignNode < BinaryNode
  def initialize(left, right)
    super(left, right)
  end

  def evaluate
    name = @left.evaluate
    result = @right.evaluate
    $calc.storevar(name, result)
    result
  end
end

#NODOS UNARIOS

class StoreNode < UnaryNode
  def initialize(sub_tree)
    super(sub_tree)
  end

  def evaluate
    $calc.store(@sub_tree.evaluate)
  end
end

class MemPlusNode < UnaryNode
  def initialize(sub_tree)
    super(sub_tree)
  end

  def evaluate
    $calc.mem_plus(@sub_tree.evaluate)
  end
end

class MemMinusNode < UnaryNode
  def initialize(sub_tree)
    super(sub_tree)
  end

  def evaluate
    $calc.mem_minus(@sub_tree.evaluate)
  end
end

#NODOS

class NumNode
  def initialize(num)
    @num = num
  end

  def evaluate
    @num
  end
end

class IdentifierNode
  def initialize(identifier)
    @identifier = identifier
  end

  def evaluate
    @identifier
  end
end

class ClearNode
  def evaluate
    $calc.clear
  end
end

class RecallNode
  def evaluate
    $calc.recall
  end
end
